src/merkle_tree/mod.rs

```diff
@@ -1,253 +1,399 @@
-        }
+    /// returns height of the merkle tree
+    pub fn height(&self) -> usize {
+        self.height
+    }
+
+    /// Returns the authentication path from leaf at `index` to root
+    pub fn generate_proof(&self, index: usize) -> Result<Path<P>, crate::Error> {
+        // gather basic tree information
+        let tree_height = tree_height(self.non_leaf_nodes.len() + self.leaf_nodes.len());
 
-        // Iterate from the leaf up to the root, storing all intermediate hash values.
-        let mut current_node = tree_index;
+        // Get Leaf hash, and leaf sibling hash,
+        let leaf_index_in_tree = convert_index_to_last_level(index, tree_height);
+        let leaf_sibling_hash = if index & 1 == 0 {
+            // leaf is left child
+            self.leaf_nodes[index + 1].clone()
+        } else {
+            // leaf is right child
+            self.leaf_nodes[index - 1].clone()
+        };
+
+        let mut path = Vec::with_capacity(tree_height - 2); // tree height - one for leaf node - one for root
+                                                            // Iterate from the bottom inner node to top, storing all intermediate hash values
+        let mut current_node = parent(leaf_index_in_tree).unwrap();
         while !is_root(current_node) {
             let sibling_node = sibling(current_node).unwrap();
-            let (curr_hash, sibling_hash) = (
-                self.tree[current_node].clone(),
-                self.tree[sibling_node].clone(),
-            );
-            if is_left_child(current_node) {
-                path.push((curr_hash, sibling_hash));
-            } else {
-                path.push((sibling_hash, curr_hash));
-            }
+            path.push(self.non_leaf_nodes[sibling_node].clone());
             current_node = parent(current_node).unwrap();
         }
 
-        // Store the root node. Set boolean as true for consistency with digest
-        // location.
-        assert!(path.len() < Self::HEIGHT as usize);
-        if path.len() != (Self::HEIGHT - 1) as usize {
-            path.push((self.tree[0].clone(), empty_hash));
-            for &(ref hash, ref sibling_hash) in &self.padding_tree {
-                path.push((hash.clone(), sibling_hash.clone()));
-            }
-        }
-        end_timer!(prove_time);
-        if path.len() != (Self::HEIGHT - 1) as usize {
-            Err(Error::IncorrectPathLength(path.len()).into())
+        debug_assert_eq!(path.len(), tree_height - 2);
+
+        // we want to make path from root to bottom
+        path.reverse();
+
+        Ok(Path {
+            leaf_index: index,
+            auth_path: path,
+            leaf_sibling_hash,
+        })
+    }
+
+    /// Given the index and new leaf, return an updated path in order from root to bottom non-leaf level
+    fn updated_path<L: ToBytes>(
+        &self,
+        index: usize,
+        new_leaf: &L,
+    ) -> Result<(LeafDigest<P>, Vec<TwoToOneDigest<P>>), crate::Error> {
```

Comment:
Should this instead return a path struct?

Reply:
I don't think so. `Path` struct is used for proof and only includes `auth_path` which is the sibling node. This `updated_path` is a helper method that generates necessary information updating the merkle tree. We don't need to update the sibling node in the process. Also, `Path` does not contain the on-path intermediate hashes output which is useful for updating the tree.

User can call `generate_proof` after updating a path to get an updated Path. `generate_proof` is quite cheap because this method does not involve any hash computation.